### PR TITLE
Add Kelly Criterion based strategies.

### PR DESCRIPTION
**Pull request by Chanka0**
# Description
Adds two new strategies:
- **KELLY**
- **KELLY_FRACTIONAL**

Both are based on the [Kelly Criterion](https://en.wikipedia.org/wiki/Kelly_criterion#Gambling_formula), a logarithmic formula used to maximize payout based on chance and odds. **KELLY** uses the returned value to place the bet, often somewhere between 20-30% of total points. **KELLY_FRACTIONAL** uses the percentage defined in BetSettings and then bets `KELLY*percentage`.

It should be stated that both strategies overrides the percentage bet. It still abides to the max point bet but will bet anything below that. This means that there is a very unlikely but potential scenario where it will bet all points. The conditions for this would need to be extraordinary (99:1 odds or similar).
## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

I've tested this implementation for the past few days on a cloned repository but added and polished it for the pull request.

Yesterday it ran on a stream with **8** completed bets with the average placed bet being **14%** of the total balance. My balance went from **19.7k** to **59.84k** by the end of the night. Now whether or not it performs better than other methods I haven't checked, but for almost 70 years after its publication it's been considered reliable and I feel my implementation lives up to that.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md)
- [x] My changes generate no new warnings
- [x] Any dependent changes have been updated in requirements.txt

### CODE REVIEW

**Comment by Chanka0** [comment]
Added 3 new strategies and renamed the 2 older ones.

- KELLY_CROWD (Was KELLY)
- KELLY_CROWD_FRACTIONAL (Was KELLY_FRACTIONAL)
- KELLY_ODDS
- KELLY_ODDS_FRACTIONAL
- KELLY_HYBRID
ODDS behaves similar to CROWD except uses the odds percentage as an indicator for win rate. I haven't tested it as much as KELLY_CROWD but it's functionally identical.

KELLY_HYBRID is basically SMART but using the Kelly Criterion to determine the bet instead of a flat percentage.

**Comment by aVitomin** [comment]
interesting

**Comment by Chanka0** [comment]
KELLY_HYBRID was removed because of compatibility issues with SMART decision making and duplicate code was consolidated into a single function.

File: TwitchChannelPointsMiner/classes/entities/Bet.py [hunk 1]
```diff
@@ -261,6 +265,13 @@ def skip(self) -> bool:
         else:
             return False, 0  # Default don't skip the bet
 
+    #Revised Kelly function to avoid duplicate code.
+    def kelly(self, indicator, fractional = False, index = 0) -> float:
```
Comment (RakSrinaNa) [inline TwitchChannelPointsMiner/classes/entities/Bet.py:269]: If `self.outcomes[index][OutcomeKeys.ODDS]` are `1` => kaboom.

**Comment by RakSrinaNa** [comment]
I'm not really sure that Kelly based on the odds is really interesting nor does anything.

Odds are simply saying that it is based on points. If you apply the Kelly formula with the values for odds/points you'd get the following.

Kelly base formula is

CodeCogsEqn (1)

With : 
* `p` your win probability (here based on ODDS_PERCENTAGE)
* `q` your loss probability, so `1-p`
* `b` your gain proportion

In this case they'd be : 
CodeCogsEqn

If you compute Kelly with these values of p, q and b you'll always get 0 (for simplicity I replace pointsA with x and pointsB with y) : 
image
https://www.wolframalpha.com/input?i=%28x%2F%28x%2By%29%29-%281-x%2F%28x%2By%29%29%2F%28%28x%2By%29%2Fx+-+1%29

**Comment by Chanka0** [comment]
> I'm not really sure that Kelly based on the odds is really interesting nor does anything.
> 
> Odds are simply saying that it is based on points. If you apply the Kelly formula with the values for odds/points you'd get the following.
> 
> Kelly base formula is
> 
> CodeCogsEqn (1)
> 
> With :
> 
>     * `p` your win probability (here based on ODDS_PERCENTAGE)
> 
>     * `q` your loss probability, so `1-p`
> 
>     * `b` your gain proportion
> 
> 
> In this case they'd be : CodeCogsEqn
> 
> If you compute Kelly with these values of p, q and b you'll always get 0 (for simplicity I replace pointsA with x and pointsB with y) : image https://www.wolframalpha.com/input?i=%28x%2F%28x%2By%29%29-%281-x%2F%28x%2By%29%29%2F%28%28x%2By%29%2Fx+-+1%29

My bad. I haven't tested the ODDS implementation nearly as much as the crowdbased implementation, in fact the only reason I added it was an attempt to introduce variety. I guess I'll revise this pull request again, backtrack it all the way to first commit. May be sometimes the first idea is the best.

**Comment by Yishaqel** [comment]
I don't recommend using this strategy. After trying it for a couple of days I ran out of points on several channels.

**Comment by Chanka0** [comment]
I advised using KELLY_FRACTIONAL in the readme.md. Since this is trying to
get optimal payout it will often place large bets against the winning side
to maximize payout, a risky but intended method. KELLY will often do 10-20%
bets, obviously a ridiculous amount if you want guaranteed points. That's
why KELLY_FRACTIONAL was created.

Personally I use KELLY_FRACTIONAL with a percentage of 50.

**Comment by Yishaqel** [comment]
I didn't read the recommended configuration correctly then

> 

So, sorry, it was my fault for not reading it correctly.

**Comment by Chanka0** [comment]
Did some backtesting, deciding to close this pull request. Strategy is too ineffective when betting on upsets.

**Pull request closed without merging**